#ASN1-8

For the final training assignment, ASN1-8, we will be diverging and doing different assignments depending on your group project. I have tried to ensure that these will all take a comparable amount of time and effort for people in each group type, but they are all pretty distinct and vary from DataCamp to homebrew assignments done on TOD-compute. 

All three assignments are pretty substantial and are designed to fill at least this week and next and to give everyone what they need to really get going on the research projects. That said, I have not given these as assignments before and we are on untrodden ground here with our whole situation. If you find this is an unrealistic amount of work for two weeks, please speak with me. Please let me know if you have any concerns.

I simply put all DataCamp courses in everyone's account. You are not responsible for doing all of them (or any of them), just the ones described in this assignment for your group type.

## Transcriptomics Groups

For groups working on transcriptomics projects, I would like you to complete two DataCamp courses, Introduction to Importing Data in R, and Data Visualization with ggplot2. For many types of extended analyses, we might like to perform similar analyses to those in ASN1-7, identifying patterns in the genes that were differentially expressed in an experiment. Instead of using premade collections of genes like in ASN1-7, we may want to run other programs and generate values for each gene, or import data from other databases or data sources. Finally, we would like to be able to visually communicate the relationships and conclusions we find in our data. We may either carefully manipulate and export the data from R to plot with spreadsheets and Excel or Sheets, or we may plot directly in R using more powerful tools like ggplot2. These two courses should give us the flexibility and ability to perform a variety of analyses, depending on the particular specific research questions we decide to ask.

You should have these assignments available to you in your DataCamp account. You may submit a document stating that you completed these courses to ELMS.

## COVID Phylogenetics Groups

For groups working on COVID phylogenetics projects, I could not find any pre-made interactive things like DataCamp that I think would help you much, so I have a similar homebrew training assignment to previous assignments. This can be found at [https://jgoodson.github.io/FIRE-TOD/ASN1-8a.html](https://jgoodson.github.io/FIRE-TOD/ASN1-8a.html)  and is a crash-course in the basics of phylogenetic analysis, building on earlier assignments conceptually. In much the same way that transcriptomics projects can take many forms with an infinite amount of specific questions, but has a shared core set of steps, evolutionary analysis has a similar series of core steps. These include multiple sequence alignment, alignment trimming, tree reconstruction, and tree visualization. All of these steps can be done with command-line tools, web-based tools, or standalone GUI programs, with decreasing levels of automation ability. Since our subject matter is rapidly changing, we may wish to update and re-run our analyses frequently. For this reason, we will focus on command-line tools for these projects.

Please submit your answers as usual.

## Comparative Genomics and Machine Learning Groups

For groups working on the comparative genomics and machine learning project, I would like you to complete a couple of DataCamp courses, the "Introduction to Python", "Intermediate Python" and "Introduction to Deep Learning in Python" courses.I will provide a video a little later this week with an overview of the whole project, and how individual group projects can fit into the overall goals. Ultimately, these will be a mix of general Python scripting to interface with biological sequence databases, transform the data, develop and run machine learning algorithms, and transform the data for the next step in the process.

For better or worse, this project will be done largely with Python scripts. Depending on your own comfort level with Python, you may skip the first "Introduction to Python" course. Please skim over this course in any case, to ensure you are comfortable with the basics. I would like you to judge your own ability here. If you are already comfortable with Python, please complete the "Intermediate Python" and "Introduction to Deep Learning in Python" courses.  If you are not comfortable with Python, for ASN1-8 please complete the "Introduction to Python" course as well. 

If you skip the beginner course (and hence are well-prepared), I would like you to complete a series of short challenges on the Rosalind Bioinformatics Challenges platform. In particular, I would like you to apply what you learned from the Python DataCamp to solving the problems in the Bioinformatics Armory section [Rosalind](http://rosalind.info/problems/tree-view/?location=bioinformatics-armory) Please complete the progression through the MEME, CLUS, and ORFR challenges (You do not need to complete the TFSQ->BFIL series challenges). You may find a page explaining how to get started using Python on the TOD-Compute server at [https://jgoodson.github.io/FIRE-TOD/ASN1-8a.html](https://jgoodson.github.io/FIRE-TOD/ASN1-8a.html) These challenges will introduce you to BioPython and interacting and parsing the types of data formats we will need for these projects.

You should have these assignments available to you in your DataCamp account. The due date for ASN1-8 will be Friday, Apr 17. You may just submit a document stating that you completed these courses and a screenshot of the Rosalind completion to ELMS.